File: LeetCode/2369. Check if There is a Valid Partition For The Array/solution.py
class Solution:
    def validPartition(self, nums: List[int]) -> bool:
        n = len(nums)
        @cache
        def dfs(i):
            if i == n:
                return True
            if i == n-1:
                return False
            if nums[i] != nums[i+1]:
                if nums[i]+1 != nums[i+1]:
                    return False
                if i >= n-2 or nums[i+1]+1 != nums[i+2]:
                    return False
                return dfs(i+3)
            else:
                if dfs(i+2):
                    return True
                if i+2 < n and nums[i+1] == nums[i+2] and dfs(i+3):
                    return True
            return False
        return dfs(0)

File: LeetCode/2369. Check if There is a Valid Partition For The Array/solution.cpp
class Solution {
public:
    bool validPartition(vector<int>& nums) {
        int n=nums.size();
        vector<bool>dp(n+1);
        dp[0]=1;
        for(int i=0;i<n;++i)
            if(dp[i]){
                if(i+1<n&&nums[i]==nums[i+1])dp[i+2]=1;
                if(i+2<n&&nums[i]==nums[i+1]&&nums[i]==nums[i+2])dp[i+3]=1;
                if(i+2<n&&nums[i]+1==nums[i+1]&&nums[i]+2==nums[i+2])dp[i+3]=1;
            }
        return dp[n];
    }
};

